File: 1 - metodosBusca/python/app/core/busca/busca.py
from abc import ABC, abstractmethod
from collections import defaultdict

from .estado import Estado
from .mostra_status_console import MostraStatusConsole
from .status import Status
from .nodo import Nodo

class Busca(ABC):
    def __init__(self, mostra_status_console=None):
        self.parar = False
        self.podar = True
        self.usar_fechado = True
        self.status = Status()
        self.mstatus = mostra_status_console
        self.fechados = None

        if mostra_status_console:
            self.set_mostra(mostra_status_console)

    def init_fechados(self):
        self.fechados = defaultdict(lambda: float('inf'))

    def get_status(self):
        return self.status

    def novo_status(self):
        self.status = Status()
        if self.mstatus:
            self.mstatus.set_status(self.status)
            self.status.set_mostra(self.mstatus)
        return self.status

    def set_mostra(self, mostra_status_console: MostraStatusConsole):
        self.mstatus = mostra_status_console
        self.mstatus.set_status(self.status)
        self.status.set_mostra(self.mstatus)

    def __str__(self):
        return "Algoritmo de busca geral"

    @abstractmethod
    def busca(self, inicial:Estado)->Nodo:
        pass

    def set_parar(self, b: bool):
        self.parar = b

    def para(self):
        self.parar = True
        self.status.termina(False)

    def set_podar(self, b):
        self.podar = b

    def usar_fechados(self, b):
        self.usar_fechado = b

    def sucessores(self, pai:Nodo)->list[Nodo]:
        return self.so_novos(pai.estado.sucessores(), pai)

    def antecessores(self, pai:Nodo)->list[Nodo]:
        try:
            return self.so_novos(pai.estado.antecessores(), pai)
        except AttributeError:
            print(f"O estado {pai.estado} nao implementa antecessores!")
            return []

    def so_novos(self, estados:list[Estado], pai:Nodo)->list[Nodo]:
        suc_nodo = []
        for e in estados:
            filho = Nodo(e, pai)
            if self.podar:
                if self.usar_fechado and self.fechados is not None:
                    custo = self.fechados[e]
                    if custo == float('inf') or filho.g < custo:
                        suc_nodo.append(filho)
                        self.fechados[e] = filho.g
                elif filho.eh_descendente_novo(pai):
                    suc_nodo.append(filho)
            else:
                suc_nodo.append(filho)
        return suc_nodo

File: 1 - metodosBusca/cSharp/Busca.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;


namespace cSharp
{
    abstract class Busca 
    {
        protected bool parar = false;
        protected bool podar = true;
        protected bool usarFechado = true;

        protected Status status = new Status(); // a classe que tem o status (model)
        protected MostraStatusConsole mstatus = null; // a classe que mostra o stauts (view)

        private Map<Estado,Integer> fechados = null; // mapeia o estado para um custo g
        
        /** busca sem mostrar status */
        public Busca() 
        {
        }

        /** busca mostrando status */
        public Busca(MostraStatusConsole ms) 
        {
            setMostra(ms);
        }
        
        protected void initFechados() 
        {
            fechados = new HashMap<Estado,Integer>();
        }
        
        public Status getStatus() 
        {
            return status;
        }
        
        public Status novoStatus() 
        {
            status = new Status();
            if (mstatus != null) 
            {
                mstatus.setStatus(status);
                status.setMostra(mstatus);
            }
            return status;
        }
        
        public void setMostra(MostraStatusConsole ms) 
        {
            mstatus = ms;
            ms.setStatus(status);
            status.setMostra(ms);        
        }
        
        public String toString() 
        {
            return "Algoritmo de busca geral";
        }
        
        public abstract Nodo busca(Estado inicial) throws Exception;
        
        public void setParar(bool b) 
        {
            parar = b;
        }
        
        public void para() 
        {
            parar = true;
            status.termina(false);
        }

        public void setPodar(bool b) 
        {
            podar = b;
        }
        
        public void usarFechados(bool b) 
        {
            usarFechado = b;
        }
        
        /**
        * gera uma lista de sucessores do nodo.
        */
        public List<Nodo> sucessores(Nodo pai) 
        {
            return soNovos(pai.estado.sucessores(),pai); // lista de todos os sucessores
        }

        public List<Nodo> antecessores(Nodo pai) 
        {
            try {
                return soNovos( ((Antecessor)pai.estado).antecessores(),pai);
            } catch (Exception e) {
                Console.WriteLine("O estado "+pai.estado+" nao implementa antecessores!");
                return new LinkedList<Nodo>();
            }
        }
        
        private List<Nodo> soNovos(List<Estado> estados, Nodo pai) 
        {
            List<Nodo> sucNodo   = new LinkedList<Nodo>(); // a lista de sucessores novos
            foreach (Estado e in estados) 
            {
                Nodo filho = new Nodo( e, pai);
                if (podar) 
                {
                    if (usarFechado && fechados != null) 
                    {
                        Integer custo = fechados.get(e);
                        if (custo == null || filho.g < custo.intValue()) 
                        { // nao esta em fechados ou tem custo menor
                            sucNodo.add(filho);
                            fechados.put(e, filho.g);
                        }
                    } else if (filho.ehDescendenteNovo(pai)) 
                    { // poda os filhos que tem um ascensor igual a ele
                        sucNodo.add(filho);
                    }
                } else 
                {
                    sucNodo.add(filho);
                }
            }
            return sucNodo;
        }
    }
}
